<div align="center">
  
  <p align="center">
    <h1>Plots and Visualizations</h1>
    <p>A collection of some materials and plots created to scientific publications</p>
  </p>
</div>

  <h3 align="center">Metagenomics and Viral reports</h3>
  <p align="center">From the paper: "Theobroma cacao Virome: Exploring Public RNA-Seq Data for Viral Discovery and Surveillance"</p>

<p align="center">Available in: https://doi.org/10.3390/v17050624</p>

<p align="center">
  <strong>Viral sequences overview</strong>
</p>
<img src="https://raw.githubusercontent.com/gabrielvpina/dataViz/refs/heads/main/figures/data_report.png" width="850" height="650">
Metavirome overview. (A) Total contigs (N) from the de novo assembly of 109 cocoa RNA-seq libraries, classified as Meta, Host, or Viral based on protein alignment. Host material refers to residual reads that did not map to the reference genome. (B) Alignment targets of viral contigs based on protein alignment. (C) Identity and coverage of contigs aligned to viral protein sequences. (D) Classification of known and novel viruses by family and genetic material, as characterized in the contigs. (E) Distribution of complete, fragmented, and absent viral conserved domains in the characterized sequences. (F) Nucleotide similarity of contigs within each characterized viral family.

<p align="center">
  <strong>Metagenomics in 109 libraries</strong>
</p>
<img src="https://raw.githubusercontent.com/gabrielvpina/dataViz/refs/heads/main/figures/metagenomics.png" width="850" height="650">
(A). Microbiome composition across the 109 Theobroma cacao-derived RNA-seq libraries, showing the relative abundance of the most frequent phyla. Phyla are represented only if they account for ≥0.05% of the total reads in each library. The X-axis indicates the relative number of reads classified within each phylum. (B). Reads abundance (%) in total metatranscriptome of the most frequent genera across all samples, considering only genera that account for ≥0.5% of the total reads in each library. 

